brain_observatory_analysis/ophys/data_formatting.py

```diff
@@ -0,0 +1,533 @@
+    trace_df = pd.DataFrame(columns=['cell_specimen_id', 'trace', 'timepoints', 'oeid', 'target_region', 'depth_order', 'bisect_layer']).set_index('cell_specimen_id')
+    for oeid in oeids:
+        temp_df = pd.DataFrame(columns=['cell_specimen_id', 'trace', 'timepoints', 'oeid', 'target_region', 'depth_order', 'bisect_layer'])
+        if trace_type == 'dff':
+            trace_df = lamf_group.experiments[oeid].dff_traces
+            trace = trace_df.dff.values
+        elif trace_type == 'events':
+            trace_df = lamf_group.experiments[oeid].events
+            trace = trace_df.events.values
+        else:
+            raise ValueError('trace_type must be either dff or events')
+        csid = trace_df.index.values
+        temp_df['cell_specimen_id'] = csid
+        temp_df['trace'] = trace
+        temp_df['timepoints'] = lamf_group.experiments[oeid].ophys_timestamps
+        temp_df['oeid'] = oeid
+        temp_df['target_region'] = lamf_group.expt_table.loc[oeid].targeted_structure
+        temp_df['depth_order'] = lamf_group.expt_table.loc[oeid].depth_order
+        temp_df['bisect_layer'] = lamf_group.expt_table.loc[oeid].bisect_layer
+        temp_df.set_index('cell_specimen_id', inplace=True)
+        temp_df.sort_index(inplace=True)
+        trace_df = pd.concat([trace_df, temp_df])
+    return trace_df
+
+
+def get_trace_df_no_task(lamf_group, session_name, trace_type='dff'):
+    """Get a dataframe of traces from a session (all ophys trace, either dff or events) with no task
+    Assume 5 min gray screen before and after task
+    and 5 min fingerprint (movie-watching; 30 sec 10 iterations) imaging at the end
+    Some sessions don't have them.
+
+    Parameters
+    ----------
+    lamf_group : ExperimentGroup
+        ExperimentGroup object
+    session_name : str
+        session name
+    trace_type : str, optional
+        'dff' or 'events', by default 'dff'
+    
+    Returns
+    -------
+    trace_df_gray_pre_task : pd.DataFrame
+        dataframe of traces from gray screen before task
+    trace_df_gray_post_task : pd.DataFrame
+        dataframe of traces from gray screen after task
+    trace_df_fingerprint : pd.DataFrame
+        dataframe of traces from fingerprint (movie-watching; 30 sec 10 iterations) imaging at the end
+    """
+    gray_period = 5 * 60  # 5 minutes
+    oeids = np.sort(lamf_group.expt_table[lamf_group.expt_table.session_name==session_name].index.values)
+    stim_df = data_formatting.annotate_stimuli(lamf_group.experiments[oeids[0]])  # First experiment represents the session stimulus presentations
+    
+    # Match the # of indices from each experiment
+    start_inds = []
+    end_inds = []
+    post_gray_end_inds = []
+    for oeid in oeids:
+        timestamps = lamf_group.experiments[oeid].ophys_timestamps
+        
+        first_stim_start_time = stim_df.start_time.values[0]
+        if first_stim_start_time > gray_period - 30:  # 30 is an arbitrary buffer
```

Comment:
what is the function of the buffer? and should it be stated more explicitly? (like have a variable name?)

Reply:
This point is somewhat temporary ducktaping. Marina told me that the new allensdk now has the table about timepoints for gray screens and fingerprints, but we cannot use it because of the long loading time. Instead, 5 minute period is estimated as the duration of each epoch. 
I have not tested it, but was concerned about some cases where the duration might be shorter than 5 minutes. So the buffer is an arbitrary safety (in seconds).
I left these in the comments + TODO to accommodate the updated allensdk in the future